File: infra/tenant/naming.ts
/**
 * Every name this component derives from a tenant slug, in one place.
 *
 * Pure string functions with no Pulumi types, so the whole set is testable
 * without an engine. That matters more here than it did on the GCP shape:
 * these names are not resource identifiers a provider would reject if wrong,
 * they are filesystem paths, Docker volume names and systemd instance names
 * on a host shared by every tenant. A collision is silent.
 */

/**
 * The slug charset, shared with `db/provision/naming.py`. Kept identical
 * deliberately: the same string becomes a MySQL account name there and a
 * systemd instance name here, and a slug that is valid in one place and not
 * the other produces a tenant that half-exists.
 */
const TENANT_SLUG_PATTERN = /^[a-z][a-z0-9-]*$/;

/** Mirrors `db/provision/naming.py`'s `TENANT_DB_PREFIX`. */
export const TENANT_DB_PREFIX = 'ghost_';

/**
 * MySQL's account-name limit is 32 characters and the tenant's database and
 * dedicated user share one name of `ghost_<sql-id>`, so the slug has 26 to
 * work with. Every other limit this slug meets is looser — systemd instance
 * names and Docker volume names are far longer, and
 * `branchleft_deploy.py`'s own stack-name pattern allows 32 — so this is the
 * binding one and the only one worth encoding.
 */
export const MAX_TENANT_SLUG_LENGTH = 32 - TENANT_DB_PREFIX.length;

/**
 * Stack names already in use on an app host by something that is not a
 * tenant.
 *
 * A tenant's Compose project name *is* its directory under `/opt/branchleft`,
 * its `/etc/branchleft/<name>.env` secrets file and its
 * `branchleft-compose@<name>` unit. Provisioning a tenant slugged `website`
 * would therefore land on top of the marketing site's stack — overwriting its
 * secrets file and its Compose project — and nothing in Docker, systemd or
 * Pulumi would object. The refusal has to be here because this component is
 * the only thing that sees the slug before anything is written.
 */
export const RESERVED_STACK_NAMES: readonly string[] = ['website', 'edge', 'db', 'monitoring'];

export function validateTenantSlug(slug: string): void {
  if (!TENANT_SLUG_PATTERN.test(slug)) {
    throw new Error(
      `GhostTenant: tenant slug "${slug}" must start with a lowercase letter and contain only ` +
        `lowercase letters, digits and hyphens.`
    );
  }
  if (slug.length > MAX_TENANT_SLUG_LENGTH) {
    throw new Error(
      `GhostTenant: tenant slug "${slug}" is ${slug.length} characters; must be at most ` +
        `${MAX_TENANT_SLUG_LENGTH} so "${TENANT_DB_PREFIX}" plus the slug fits MySQL's ` +
        `32-character account-name limit.`
    );
  }
  if (RESERVED_STACK_NAMES.includes(slug)) {
    throw new Error(
      `GhostTenant: tenant slug "${slug}" is reserved — an app host already runs a Compose ` +
        `stack of that name, and a tenant using it would overwrite that stack's directory, ` +
        `secrets file and systemd unit. Reserved: ${RESERVED_STACK_NAMES.join(', ')}.`
    );
  }
}

/** MySQL identifiers cannot carry the hyphens a slug may. */
export function sqlIdentifier(slug: string): string {
  return slug.replaceAll('-', '_');
}

/** The tenant's logical database and its dedicated DB user share this name. */
export function databaseAndUserName(slug: string): string {
  return `${TENANT_DB_PREFIX}${sqlIdentifier(slug)}`;
}

/**
 * The Compose project name, which is also the systemd instance name, the
 * directory under `/opt/branchleft` and the stem of both files under
 * `/etc/branchleft`. One value, because `branchleft-compose@.service` and
 * `branchleft-deploy` already treat them as one.
 */
export function stackName(slug: string): string {
  return slug;
}

export function stackDirectory(slug: string): string {
  return `/opt/branchleft/${stackName(slug)}`;
}

export function composeUnitName(slug: string): string {
  return `branchleft-compose@${stackName(slug)}.service`;
}

/** Root-owned `0600`, written by an operator and by no automated path. */
export function secretsEnvPath(slug: string): string {
  return `/etc/branchleft/${stackName(slug)}.env`;
}

/** Written by `branchleft-deploy` alone; never the same file as the above. */
export function imageEnvPath(slug: string): string {
  return `/etc/branchleft/${stackName(slug)}.image.env`;
}

/**
 * Volume names are given explicitly in the rendered Compose file rather than
 * left to Compose's `<project>_<volume>` prefixing.
 *
 * The host-side provisioning step has to create these volumes, own them to
 * the tenant UID and refuse a UID another tenant already holds, and it runs
 * before any Compose project exists to derive a prefix from. A name Compose
 * would have synthesised is a name that step would have to reconstruct from
 * knowledge of Compose's prefixing rule — so it is stated once, here, and
 * both sides read it.
 */
export function contentVolumeName(slug: string): string {
  return `ghost-${slug}-content`;
}

export function adaptersVolumeName(slug: string): string {
  return `ghost-${slug}-adapters`;
}

File: db/provision/naming.py
#!/usr/bin/env python3
"""Every name this stack derives from a tenant name, in one place.

The charset rule mirrors `infra/tenant/naming.ts`'s `validateTenantSlug`,
reused rather than re-derived -- kept identical deliberately, because the same
string becomes a MySQL account name here and a systemd instance name there,
and a name valid on one side and not the other produces a tenant that
half-exists. The length limit does not mirror it: `infra/tenant/naming.ts`'s
26-character bound and this module's are both derived from MySQL's own
32-character account-name limit independently, so they agree by
construction rather than by copying a number.

The trailing character is restricted to a letter or digit for the same reason
it is on the TypeScript side: `infra/tenant/media.ts`'s `mediaBucketName`
turns this same slug into an S3-compatible bucket name, and S3 bucket naming
rules require a bucket name to both start and end with a lowercase letter or
digit.

MySQL's own account name limit -- 32 characters, unchanged since 5.7.8 and
confirmed current in the MySQL 8.0 reference manual
(https://dev.mysql.com/doc/refman/8.0/en/user-names.html) -- applies to
`TENANT_DB_PREFIX + tenant_name`, since that combined string is both the
database name and the account name this stack creates.
"""

from __future__ import annotations

import re

TENANT_DB_PREFIX = "ghost_"

MAX_MYSQL_ACCOUNT_NAME_LENGTH = 32

MAX_TENANT_NAME_LENGTH = MAX_MYSQL_ACCOUNT_NAME_LENGTH - len(TENANT_DB_PREFIX)

TENANT_NAME_PATTERN = re.compile(r"\A[a-z]([a-z0-9-]*[a-z0-9])?\Z")

# The tenant's dedicated DB user only ever connects from the app hosts, over
# the private subnet -- scoping the account's host part to it is a second,
# independent boundary alongside `require_secure_transport`.
TENANT_USER_HOST = "10.20.1.%"


class InvalidTenantName(ValueError):
    """Raised for a tenant name this stack refuses to provision from."""


def validate_tenant_name(tenant_name: str) -> None:
    if not TENANT_NAME_PATTERN.match(tenant_name):
        raise InvalidTenantName(
            f"tenant name {tenant_name!r} must start with a lowercase letter, end with a "
            "lowercase letter or digit, and contain only lowercase letters, digits and "
            "hyphens in between"
        )
    if len(tenant_name) > MAX_TENANT_NAME_LENGTH:
        raise InvalidTenantName(
            f"tenant name {tenant_name!r} is {len(tenant_name)} characters; must be at "
            f"most {MAX_TENANT_NAME_LENGTH} so \"{TENANT_DB_PREFIX}\" plus the name fits "
            f"MySQL's {MAX_MYSQL_ACCOUNT_NAME_LENGTH}-character account name limit"
        )


def sql_identifier(tenant_name: str) -> str:
    """MySQL identifiers cannot carry the hyphens a tenant name may."""
    return tenant_name.replace("-", "_")


def database_and_user_name(sql_id: str) -> str:
    """The tenant's logical database and its dedicated DB user share this name."""
    return f"{TENANT_DB_PREFIX}{sql_id}"
